compiler/rustc_driver_impl/src/args.rs

```diff
@@ -5,39 +5,97 @@ use std::io;
 
 use rustc_session::EarlyErrorHandler;
 
-fn arg_expand(arg: String) -> Result<Vec<String>, Error> {
-    if let Some(path) = arg.strip_prefix('@') {
-        let file = match fs::read_to_string(path) {
-            Ok(file) => file,
-            Err(ref err) if err.kind() == io::ErrorKind::InvalidData => {
-                return Err(Error::Utf8Error(Some(path.to_string())));
+/// Expands argfiles in command line arguments.
+#[derive(Default)]
+struct Expander {
+    shell_argfiles: bool,
+    next_is_unstable_option: bool,
+    expanded: Vec<String>,
+}
+
+impl Expander {
+    /// Handles the next argument. If the argument is an argfile, it is expanded
+    /// inline.
+    fn arg(&mut self, arg: &str) -> Result<(), Error> {
+        if let Some(argfile) = arg.strip_prefix('@') {
+            match argfile.split_once(':') {
+                Some(("shell", path)) if self.shell_argfiles => {
+                    shlex::split(&Self::read_file(path)?)
+                        .ok_or_else(|| Error::ShellParseError(path.to_string()))?
+                        .into_iter()
+                        .for_each(|arg| self.push(arg));
+                }
```

Comment:
What about `@other_string:path`?

Reply:
That should fall into the second case because `"other_string" != "shell"`. I assume that if `shell_argfiles` is disabled or the argfile "type" specifier isn't `@shell:`, then we want to treat everything after the `@` as the path to the argfile. So it should try to open `other_string:path` as a file and read newline-separated args from it.